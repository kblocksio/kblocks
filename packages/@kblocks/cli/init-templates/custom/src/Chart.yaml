apiVersion: v2
name: <%= spec.definition.singular %>-block
description: <%= spec.definition.description %>
type: application
version: 0.1.0

# dependencies:
#   - name: postgresql
#     repository: https://charts.bitnami.com/bitnami
#     version: 15.5.20
