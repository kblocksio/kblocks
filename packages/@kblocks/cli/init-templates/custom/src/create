#!/usr/bin/env node
const fs = require("fs");

const object = JSON.parse(fs.readFileSync(process.env.KBLOCKS_OBJECT, "utf8"));

console.log("creating resource with value:", object.myValue);

fs.writeFileSync(process.env.KBLOCKS_OUTPUTS, JSON.stringify({
  myOutput: `create ${object.myValue}`,
}));
