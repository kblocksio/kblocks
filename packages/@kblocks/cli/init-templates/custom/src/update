#!/usr/bin/env node
const fs = require("fs");

const object = JSON.parse(fs.readFileSync(process.env.KBLOCKS_OBJECT, "utf8"));

console.log("updating resource with value:", object.myValue);

fs.writeFileSync(process.env.KBLOCKS_OUTPUTS, JSON.stringify({
  myOutput: `update ${object.myValue}`,
}));
