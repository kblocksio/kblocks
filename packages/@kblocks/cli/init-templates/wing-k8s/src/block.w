bring "cdk8s-plus-30" as k8s;

pub struct <%= spec.definition.kind %>Spec {
  myInput: str;
}

pub class <%= spec.definition.kind %> {
  pub myOutput: str;

  new(spec: <%= spec.definition.kind %>Spec) {

    let configMap = new k8s.ConfigMap(
      data: {
        "my-key": spec.myInput,
      }
    );

    this.myOutput = configMap.name;
  }
}