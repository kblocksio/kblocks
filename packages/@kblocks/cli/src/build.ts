import path from "path";
import fs from "fs";
import { App, Chart } from "cdk8s";
import { Operator } from "./operator";
import { Worker } from "./worker";
import { CustomResourceDefinition } from "./crd";
import { JsonSchemaProps } from "../imports/k8s";
import { ConfigMapFromDirectory } from "./configmap";
import packageJson from "../package.json";
import { Control } from "./control";
import { formatBlockTypeForEnv, IncludeManifest, Manifest } from "@kblocks/api";
import { Construct } from "constructs";
import yaml from 'yaml';
import { getManifest, renderStatusSchema } from "./manifest-util";
import { BlockRequest } from "./types";
import cloneDeep from "lodash.clonedeep";

export const buildCommand = async (opts: {
  DIR?: string;
  manifest: string;
  output: string;
  env: Record<string, string>;
}) => {
  const { blockRequests, outdir, names } = await build({
    ...opts,
    dir: opts.DIR ?? process.cwd(),
  });

  console.log();
  console.log("-------------------------------------------------------------------------------------------------------------------");
  for (const request of blockRequests) {
    console.log(`Block '${request.block.definition.group}.${request.block.definition.kind}' is ready. To install:`);
  }
  console.log();
  console.log(`  helm upgrade --install ${names}-block ${outdir}`);
  console.log();
};

export async function build(opts: {
  manifest: string;
  dir: string;
  output: string;
  silent?: boolean;
  env: Record<string, string>;
}) {
  const outdir = path.resolve(process.cwd(),opts.output);

  // create the output directory and clean it up if it already exists
  fs.rmSync(outdir, { recursive: true, force: true });
  fs.mkdirSync(outdir, { recursive: true });

  const dir = path.resolve(process.cwd(), opts.dir);
  const { manifest, additionalObjects, tmpSrc } = await getManifest({
    dir: path.resolve(process.cwd(), opts.dir),
    manifest: opts.manifest,
    outdir
  });

  // Check if Chart.yaml exists in the manifest directory
  const chartPath = path.join(dir, 'Chart.yaml');
  if (fs.existsSync(chartPath)) {
    fs.copyFileSync(chartPath, path.join(outdir, 'Chart.yaml'));
  }

  const blockRequests: BlockRequest[] = [];
  const moreObjects: any[] = [...additionalObjects];
  for (const include of manifest.include ?? []) {
    const includePath = path.resolve(dir, include);
    const included = await getManifest({
      dir: path.dirname(includePath),
      manifest: path.basename(includePath),
      outdir
    });

    blockRequests.push({
      block: included.manifest,
      source: included.tmpSrc,
    });
    moreObjects.push(...included.additionalObjects);
  }

  const names = synth({
    mainBlock: { block: manifest, source: tmpSrc },
    included: blockRequests,
    output: outdir,
    env: opts.env,
  });

  // write any additional objects to the templates directory
  if (moreObjects.length > 0) {
    const additionalObjectsManifest = path.join(outdir, "templates", "additional-objects.yaml");
    fs.writeFileSync(additionalObjectsManifest, yaml.stringify(moreObjects));
  }

  return { outdir, names, blockRequests };
}

export interface BlockProps {
  env: Record<string, string>;
  blockRequests: BlockRequest[];
}

interface SynthProps {
  mainBlock: BlockRequest;
  included: BlockRequest[];
  env: Record<string, string>;
  output: string;
  force?: boolean;
}

export class Block extends Chart {
  constructor(scope: Construct, id: string, props: BlockProps) {
    super(scope, id);

    const { env } = props;

    // make a deep copy of the block requests so we can modify them
    const blockRequests: BlockRequest[] = cloneDeep(props.blockRequests);

    if (blockRequests.length === 0) {
      throw new Error("No block requests provided");
    }

    // add the "status" schema to the all blocks
    for (const req of blockRequests) {
      req.block.definition.schema.properties = req.block.definition.schema.properties ?? {};
      req.block.definition.schema.properties.status = renderStatusSchema(req.block);
    }

    const mainBlock = blockRequests[0];
    const namespace = mainBlock.block.operator?.namespace ?? "{{ .Release.Namespace }}";
    const workers = mainBlock.block.operator?.workers ?? 1;
    const flushOnly = mainBlock.block.operator?.flushOnly ?? false;

    addSystemIfNotSet(mainBlock.block);

    const configmap = new ConfigMapFromDirectory(this, "ConfigMapVolume", {
      blockRequests,
      namespace,
      flushOnly,
    });
  
    const redisServiceName = `${id}-redis`;

    const image = {
      operator: calculateImageName("operator"),
      worker: calculateImageName("worker"),
      control: calculateImageName("control"),
    };

    const pod = {
      namespace,
      configMaps: configmap.configMaps,
      ...mainBlock.block.operator,
      env: {
        REDIS_URL: `redis://${redisServiceName}.${namespace}.svc.cluster.local:${6379}`,
        // this can be used to spawn new blocks (e.g. by the `Block` block).
        KBLOCKS_WORKER_IMAGE: image.worker,
        KBLOCKS_OPERATOR_IMAGE: image.operator,
        KBLOCKS_CONTROL_IMAGE: image.control,
        ...mainBlock.block.operator?.env,
        ...env,
      }
    }
    const blocks = blockRequests.map(b => ({
      ...b.block.definition,
    }));
  
    new Operator(this, "Operator", {
      names: id,
      namespace,
      image: image.operator,
      redisServiceName,
      workers,
      blocks,
      pod,
    });
  
    if (!flushOnly) {
      new Worker(this, "Worker", {
        names: id,
        namespace,
        image: image.worker,
        replicas: workers,
        blocks,
        pod,
      });
    }
  
    new Control(this, "Control", {
      names: id,
      namespace,
      image: image.control,
      workers,
      blocks,
      pod,
    });

    if (!mainBlock.block.operator?.skipCrd) {
      for (const blockRequest of blockRequests) {
        const annotations: Record<string, string> = {
          "kblocks.io/metadata-namespace": namespace,
        };
      
        if (blockRequest.block.definition.icon) {
          annotations["kblocks.io/icon"] = blockRequest.block.definition.icon;
        }
      
        if (blockRequest.block.definition.color) {
          annotations["kblocks.io/color"] = blockRequest.block.definition.color;
        }
    
        const blockType = formatBlockTypeForEnv(blockRequest.block.definition);

        new CustomResourceDefinition(this, `CRD-${blockType}`, {
          ...blockRequest.block.definition,
          annotations,
          schema: blockRequest.block.definition.schema,
        });
      }
    }
  }
}

export function synth(opts: SynthProps) {
  const { blockRequests } = validateBlockRequests(opts.mainBlock, opts.included);
  const names = calculateNames(blockRequests);
  const outputDir = path.resolve(opts.output);
  
  const templatesdir = path.join(outputDir, "templates");
  fs.mkdirSync(templatesdir, { recursive: true });

  const app = new App({ outdir: templatesdir });

  new Block(app, names, {
    ...opts,
    blockRequests,
  });

  app.synth();

  const chartFilePath = path.join(outputDir, 'Chart.yaml');
  if (!fs.existsSync(chartFilePath)) {
    fs.writeFileSync(chartFilePath, yaml.stringify({
      apiVersion: 'v1',
      name: `${names}-kblock`,
      version: "1.0.0"
    }));
  }

  return names;
}

function resolveSchema(props: Manifest): JsonSchemaProps {
  if (!props.definition.schema) {
    throw new Error("No schema found in kblock manifest. Please define the OpenAPIV3 schema under the 'schema' field of the 'definition' section of the manifest.");
  }

  return props.definition.schema;
}

/**
 * Calculate the image name for a given service.
 * 
 * @param serviceName The name of the service to calculate the image name for (e.g. `operator`, `worker`, or `control`)
 * @returns The image name.
 */
function calculateImageName(serviceName: string) {
  const imagePrefix = "wingcloudbot/kblocks-";
  const versionOverrideEnv = `KBLOCKS_VERSION`;
  const overrideEnv = `KBLOCKS_${serviceName.toUpperCase()}_IMAGE`;

  // if we have an explicit override KBLOCKS_XXX_IMAGE, then use it
  const imageOverride = process.env[overrideEnv];
  if (imageOverride) {
    console.log(`Override for ${serviceName} is ${imageOverride}`);
    return imageOverride;
  }

  // if we only have a version override, use it
  const versionOverride = process.env[versionOverrideEnv];
  if (versionOverride) {
    const image = `${imagePrefix}${serviceName}:${versionOverride}`;
    return image;
  }

  // now, if this is a source build, we got to bail
  if (packageJson.version === "0.0.1") {
    throw new Error(`Building from source and neither '${overrideEnv}' nor '${versionOverrideEnv}' are set, please set one of them to the image you want to use (e.g. '${overrideEnv}=wingcloudbot/kblocks-${serviceName}:0.1.13' or '${versionOverrideEnv}=0.1.13')`);
  }

  return `${imagePrefix}${serviceName}:${packageJson.version}`;
}

function addSystemIfNotSet(block: Manifest) {
  function setEnv(key: string, optional: boolean) {
    block.operator = block.operator ?? {};
    // check if one of the "env" is KBLOCKS_API_KEY
    if (key in (block.operator.env ?? {}) ||
        key in (block.operator.envSecrets ?? {}) ||
        key in (block.operator.envConfigMaps ?? {})) {
      return;
    }
  
    block.operator.envSecrets = block.operator.envSecrets ?? {};
    block.operator.envSecrets[key] = {
      key: key,
      secret: "kblocks",
      optional: optional,
    };
  }

  const keys = ["KBLOCKS_SYSTEM_ID", "KBLOCKS_API_KEY", "KBLOCKS_PUBSUB_HOST"];
  for (const key of keys) {
    setEnv(key, false);
  }

  const optionalKeys = ["KBLOCKS_PUBSUB_PORT", "KBLOCKS_STORAGE_PREFIX"];
  for (const key of optionalKeys) {
    setEnv(key, true);
  }
}

export function calculateNames(blockRequests: BlockRequest[]) {
  if (blockRequests.length === 0) {
    throw new Error("No block requests provided");
  }

  const names = [];
  for (const { block } of blockRequests) {
    names.push(block.definition.kind.toLocaleLowerCase());
  }

  return names.join("-");
}

function validateManifest(block: Manifest) {
  try {
    Manifest.parse(block);

    if (block.source?.directory && !block.source.directory.endsWith("/src")) {
      throw new Error(`Source directory must end with '/src'. got: ${block.source.directory}`);
    }

  } catch (err: any) {
    if (typeof(err["format"]) === "function") {
      throw new Error(`Invalid block manifest: ${JSON.stringify(err.format(), null, 2)}`);
    }

    throw err;
  }
}

function validateBlockRequests(mainBlock: BlockRequest, included: BlockRequest[]) {
  if (included.length > 0) {
    const allowdProps = ["include", "operator"];
    for (const key of Object.keys(mainBlock.block)) {
      if (!allowdProps.includes(key)) {
        throw new Error(`Invalid property '${key}' in main block manifest`);
      }
    }

    IncludeManifest.parse(mainBlock.block);

    for (const { block } of included) {
      validateManifest(block);

      if (block.operator) {
        throw new Error("Included blocks cannot have an operator section");
      }
    }

    return {
      blockRequests: included.map(b => ({
        block: {
          ...b.block,
          operator: mainBlock.block.operator,
        },
        source: b.source,
      })),
    }
  } else {
    validateManifest(mainBlock.block);
    return {
      blockRequests: [mainBlock],
    }
  }
}
