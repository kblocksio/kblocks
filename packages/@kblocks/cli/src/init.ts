import fs from "fs";
import path from "path";
import yaml from "yaml";
import ejs from "ejs";
import { ApiObject, Manifest } from "./api";
import { importCommand } from "./import";

export interface InitOptions {
  TEMPLATE?: string;
  DIR?: string;
  import?: boolean;

  group?: string;
  kind?: string;
  plural?: string;
  icon?: string;
  description?: string;
  apiVersion?: string;
  color?: string;
  listKind?: string;
  shortNames?: string[];
  categories?: string[];
  singular?: string;
}

export async function initCommand(argv: InitOptions) {
  if (!argv.TEMPLATE) {
    throw new Error("Template is required. Use --template <template-name> to specify the template to use. Available templates: " + Object.keys(listProjectTemplates()).join(", "));
  }

  if (!argv.DIR) {
    throw new Error("Directory is required");
  }

  // if the directory exists, error out
  if (fs.existsSync(argv.DIR)) {
    throw new Error(`Directory '${argv.DIR}' already exists`);
  }

  const templateDir = path.join(__dirname, `../init-templates/${argv.TEMPLATE}`);
  if (!fs.existsSync(templateDir)) {
    throw new Error(`Template '${argv.TEMPLATE}' not found`);
  }

  const targetDir = path.resolve(process.cwd(), argv.DIR);


  // read the manifest from the template and apply any command line options to it so it is possible
  // to pass the manifest to the template renderer

  const docs = yaml.parseAllDocuments(fs.readFileSync(path.join(templateDir, "kblock.yaml"), "utf-8"));
  if (docs.length !== 1) {
    throw new Error("Expected exactly one YAML document in kblock.yaml");
  }

  const manifest = docs[0].toJSON() as ApiObject & { spec: Manifest };
  if (manifest.apiVersion !== "kblocks.io/v1" || manifest.kind !== "Block") {
    throw new Error("Expected the first document in kblock.yaml to be a Block");
  }

  const options = ["group", "apiVersion", "kind", "plural", "icon", "description", "color", "listKind", "shortNames", "categories", "singular"];
  for (const option of options) {
    const arg = (argv as any)[option];
    if (arg) {
      const targetOptions = option === "apiVersion" ? "version" : option;
      (manifest.spec.definition as any)[targetOptions] = arg;
    }
  }

  const errors = [];

  if (!manifest.spec.definition.group) {
    errors.push("--group is required");
  }

  if (!manifest.spec.definition.plural) {
    errors.push("--plural is required");
  }

  if (!manifest.spec.definition.kind) {
    errors.push("--kind is required");
  }

  if (!manifest.spec.definition.version) {
    errors.push("--apiVersion is required");
  }

  if (errors.length > 0) {
    throw new Error(errors.join("\n"));
  }

  manifest.spec.definition.singular = manifest.spec.definition.singular ?? manifest.spec.definition.kind.toLowerCase();

  manifest.metadata = manifest.metadata ?? {};
  manifest.metadata.name = `${manifest.spec.definition.plural}.${manifest.spec.definition.group}`;


  fs.mkdirSync(targetDir, { recursive: true });
  pasteTemplate(templateDir, targetDir, manifest);

  if (argv.import) {
    await importCommand({
      DIR: path.join(targetDir, "src"),
      group: manifest.spec.definition.group,
      apiVersion: manifest.spec.definition.version,
      kind: manifest.spec.definition.kind,
    });

    if (!manifest.spec.operator) {
      manifest.spec.operator = {};
    }
    manifest.spec.operator.skipCrd = true;
    manifest.spec.operator.flushOnly = true;
  }

  // write the kblock.yaml file
  fs.writeFileSync(path.join(targetDir, "kblock.yaml"), yaml.stringify(manifest));

  const kb = path.basename(process.argv[1]);

  console.log(`ðŸŽ‰ Your new and shiny ${manifest.spec.engine} block is ready!`);
  console.log("");
  console.log("Next steps:");
  console.log("");
  console.log("1. Add your code to the `src` directory");
  console.log(`2. Update '${manifest.spec.definition.schema}' with the input schema for your block`);
  console.log("3. Update the `kblock.yaml` file with the outputs for your block");
  console.log("");
  console.log("Once you are ready, you can run `kb enrich` in order to automatically update your README file,");
  console.log("come up with a description for the block, your schema fields and find a good icon.");
  console.log("");
  console.log(`  ${kb} enrich ${argv.DIR}`);
  console.log("");
  console.log("To install it to a cluster use:");
  console.log("");
  console.log(`  ${kb} install ${argv.DIR}`);
  console.log("");
}

export function catalogCommand() {
  const templates = listProjectTemplates();
  console.log(JSON.stringify(templates, null, 2));
}

export const listProjectTemplates = () => {
  const basedir = path.join(__dirname, "../init-templates");
  const dirs = fs.readdirSync(basedir);
  const results: Record<string, ProjectTemplate> = {};

  for (const dir of dirs) {
    // skip non-directories
    if (!fs.statSync(path.join(basedir, dir)).isDirectory()) {
      continue;
    }

    const t = readProjectTemplate(basedir, dir);
    results[dir] = t;
  }

  return results;
}

export const readProjectTemplate = (basedir: string, dir: string) => {
  const manifestFile = path.join(basedir, dir, ".kbinit", "manifest.yaml"); 
  if (!fs.existsSync(manifestFile)) {
    throw new Error(`Unable to find ${manifestFile}`);
  }

  const docs = yaml.parseAllDocuments(fs.readFileSync(manifestFile, "utf-8"));
  if (docs.length !== 1) {
    throw new Error(`Expected exactly one YAML document in kb-init.yaml for ${dir}`);
  }

  const manifest = docs[0].toJSON() as ProjectTemplate;

  const readme = fs.readFileSync(path.join(basedir, dir, "README.md"), "utf-8");

  return {
    ...manifest,
    readme,
    icon: fs.readFileSync(path.join(path.dirname(manifestFile), manifest.icon), "utf-8"),
  };
}

type ProjectTemplate = {
  name: string;
  description: string;
  icon: string;
  readme: string;
};

/**
 * Copy all files (recursively) from the template directory to the target directory, while rendering the files
 * with EJS so that the variables in the block manifest can be replaced.
 * 
 * @param templateDir  The source directory of the template 
 * @param targetDir  The target directory to copy the template to
 * @param blockManifest  The block manifest to use when rendering the template
 */
function pasteTemplate(templateDir: string, targetDir: string, blockManifest: ApiObject & { spec: Manifest }) {
  const files = fs.readdirSync(templateDir);
  for (const file of files) {
    const filePath = path.join(templateDir, file);
    const targetPath = path.join(targetDir, file);
    const stat = fs.statSync(filePath);
    if (stat.isDirectory()) {
      if (file === ".kbinit") {
        continue;
      }

      fs.mkdirSync(targetPath, { recursive: true });
      pasteTemplate(filePath, targetPath, blockManifest);
    } else {  
      const contents = fs.readFileSync(filePath, "utf-8");  
      const rendered = ejs.render(contents, blockManifest);
      fs.writeFileSync(targetPath, rendered);
    }
  }
}