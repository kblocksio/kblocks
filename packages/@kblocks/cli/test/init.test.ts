import { listProjectTemplates, initCommand, InitOptions } from "../src/init";
import { expect, test } from "vitest";
import os from "os";
import path from "path";
import fs from "fs";

const catalog = listProjectTemplates();

Object.keys(catalog).forEach((name) => {
  test(`${name} template`, async () => {
    const template = catalog[name];
    expect(template).toBeDefined();
    expect(template.name).toBeDefined();
    expect(template.description).toBeDefined();
    expect(template.readme).toBeDefined();
    expect(template.icon).toBeDefined();

    const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), name));
    console.log(tmpdir);

    fs.rmdirSync(tmpdir, { recursive: true });

    const options: InitOptions = {
      DIR: tmpdir,
      apiVersion: "v99",
      categories: ["cat1", "cat2"],
      color: "red",
      kind: "Bang",
      group: "bang.com",
      description: "the bang bang description",
      icon: "ðŸ”¥",
      listKind: "BangList",
      plural: "bangs",
      shortNames: ["b", "ba"],
      TEMPLATE: name,
    };

    if (name === "noop") {
      options.import = true;
      options.kind = "Secret";
      options.group = "core";
      options.plural = "secrets";
      options.apiVersion = "v1";
    }

    await initCommand(options);

    const files = readAllFiles(tmpdir);
    expect(files).toMatchSnapshot();
  });
});


/**
 * Returns a map of all files in a directory and its subdirectories
 * @param dir - The directory to read
 * @returns A map of all files in the directory and its subdirectories
 */
function readAllFiles(basedir: string) {
  const fileMap: Record<string, string> = {};

  const readDir = (dir: string) => {
    const files = fs.readdirSync(path.join(basedir, dir));
    for (const file of files) {
      const relPath = path.join(dir, file);
      const filePath = path.join(basedir, relPath);

      // if the file is a directory, read it recursively
      if (fs.statSync(filePath).isDirectory()) {
        readDir(relPath);
        continue;
      }

      const fileContent = fs.readFileSync(filePath, "utf8");
      fileMap[relPath] = fileContent;
    }
  };

  readDir(".");

  return fileMap;
}
