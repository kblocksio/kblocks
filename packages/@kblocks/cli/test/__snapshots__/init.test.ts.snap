// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom template 1`] = `
{
  "README.md": "# Bang Block

TODO: Add a description of the block or run \`kb enrich\` to generate one.",
  "kblock.yaml": "apiVersion: kblocks.io/v1
kind: Block
spec:
  engine: custom
  definition:
    outputs:
      - myOutput
    readme: ./README.md
    schema: ./src/values.schema.json
    description: the bang bang description
    group: bang.com
    version: v99
    kind: Bang
    plural: bangs
    icon: ðŸ”¥
    color: red
    listKind: BangList
    shortNames:
      - b
      - ba
    categories:
      - cat1
      - cat2
    singular: bang
metadata:
  name: bangs.bang.com
",
  "src/.gitignore": "charts/",
  "src/Chart.yaml": "apiVersion: v2
name: bang-block
description: the bang bang description
type: application
version: 0.1.0

# dependencies:
#   - name: postgresql
#     repository: https://charts.bitnami.com/bitnami
#     version: 15.5.20
",
  "src/create": "#!/usr/bin/env node
const fs = require("fs");

const object = JSON.parse(fs.readFileSync(process.env.KBLOCKS_OBJECT, "utf8"));

console.log("creating resource with value:", object.myValue);

fs.writeFileSync(process.env.KBLOCKS_OUTPUTS, JSON.stringify({
  myOutput: \`create \${object.myValue}\`,
}));
",
  "src/delete": "#!/usr/bin/env node
const fs = require("fs");

const object = JSON.parse(fs.readFileSync(process.env.KBLOCKS_OBJECT, "utf8"));

console.log("deleting resource", object);
",
  "src/update": "#!/usr/bin/env node
const fs = require("fs");

const object = JSON.parse(fs.readFileSync(process.env.KBLOCKS_OBJECT, "utf8"));

console.log("updating resource with value:", object.myValue);

fs.writeFileSync(process.env.KBLOCKS_OUTPUTS, JSON.stringify({
  myOutput: \`update \${object.myValue}\`,
}));
",
  "src/values.schema.json": "{
  "type": "object",
  "required": [],
  "properties": {
    "myValue": {
      "type": "string",
      "description": "An example of an input variable that will be used in the block"
    }
  }
}",
}
`;

exports[`helm template 1`] = `
{
  "README.md": "# Bang Block

TODO: Add a description of the block or run \`kb enrich\` to generate one.",
  "kblock.yaml": "apiVersion: kblocks.io/v1
kind: Block
spec:
  engine: helm
  definition:
    outputs:
      - myOutput
    readme: ./README.md
    schema: ./src/values.schema.json
    description: the bang bang description
    group: bang.com
    version: v99
    kind: Bang
    plural: bangs
    icon: ðŸ”¥
    color: red
    listKind: BangList
    shortNames:
      - b
      - ba
    categories:
      - cat1
      - cat2
    singular: bang
metadata:
  name: bangs.bang.com
",
  "src/.gitignore": "charts/",
  "src/Chart.yaml": "apiVersion: v2
name: bang-block
description: the bang bang description
type: application
version: 0.1.0

# dependencies:
#   - name: postgresql
#     repository: https://charts.bitnami.com/bitnami
#     version: 15.5.20
",
  "src/templates/NOTES.txt": "{
  "myOutput": "{{ .Release.Name }}-is-the-awesome-output-for-bang"
}",
  "src/values.schema.json": "{
  "type": "object",
  "required": [],
  "properties": {
    "myValue": {
      "type": "string",
      "description": "An example of an input variable that will be used in the block"
    }
  }
}",
}
`;

exports[`noop template 1`] = `
{
  "README.md": "# Secret Block

TODO: Add a description of the block or run \`kb enrich\` to generate one.",
  "kblock.yaml": "apiVersion: kblocks.io/v1
kind: Block
spec:
  engine: noop
  definition:
    readme: ./README.md
    schema: ./src/values.schema.json
    description: the bang bang description
    group: core
    version: v1
    kind: Secret
    plural: secrets
    icon: ðŸ”¥
    color: red
    listKind: BangList
    shortNames:
      - b
      - ba
    categories:
      - cat1
      - cat2
    singular: secret
  operator:
    skipCrd: true
    flushOnly: true
metadata:
  name: secrets.core
",
  "src/.gitignore": "charts/",
  "src/values.schema.json": "{
  "description": "Secret holds secret data of a certain type. The total bytes of the values in the Data field must be less than MaxSecretSize bytes.",
  "type": "object",
  "properties": {
    "data": {
      "description": "Data contains the secret data. Each key must consist of alphanumeric characters, '-', '_' or '.'. The serialized form of the secret data is a base64 encoded string, representing the arbitrary (possibly non-string) data value here. Described in https://tools.ietf.org/html/rfc4648#section-4",
      "type": "object",
      "additionalProperties": {
        "type": "string",
        "format": "byte"
      }
    },
    "immutable": {
      "description": "Immutable, if set to true, ensures that data stored in the Secret cannot be updated (only object metadata can be modified). If not set to true, the field can be modified at any time. Defaulted to nil.",
      "type": "boolean"
    },
    "stringData": {
      "description": "stringData allows specifying non-binary secret data in string form. It is provided as a write-only input field for convenience. All keys and values are merged into the data field on write, overwriting any existing values. The stringData field is never output when reading from the API.",
      "type": "object",
      "additionalProperties": {
        "type": "string"
      }
    },
    "type": {
      "description": "Used to facilitate programmatic handling of secret data. More info: https://kubernetes.io/docs/concepts/configuration/secret/#secret-types",
      "type": "string"
    }
  },
  "x-kubernetes-group-version-kind": [
    {
      "group": "",
      "kind": "Secret",
      "version": "v1"
    }
  ]
}",
}
`;

exports[`terraform template 1`] = `
{
  "README.md": "# Bang Block

TODO: Add a description of the block or run \`kb enrich\` to generate one.",
  "kblock.yaml": "apiVersion: kblocks.io/v1
kind: Block
spec:
  engine: tofu
  definition:
    description: the bang bang description
    icon: ðŸ”¥
    readme: ./README.md
    schema: src/values.schema.json
    outputs:
      - myOutput
    group: bang.com
    version: v99
    kind: Bang
    plural: bangs
    color: red
    listKind: BangList
    shortNames:
      - b
      - ba
    categories:
      - cat1
      - cat2
    singular: bang
  operator:
    envSecrets:
      AWS_DEFAULT_REGION: aws-credentials
      AWS_ACCESS_KEY_ID: aws-credentials
      AWS_SECRET_ACCESS_KEY: aws-credentials
metadata:
  name: bangs.bang.com
",
  "src/main.tf": "# This is an example of an AWS SNS topic resource which will be created by the block
resource "aws_sns_topic" "my_topic" {
  name = var.myValue
}

# This is an example of an output value that will be returned by the block
output "myOutput" {
  value       = aws_sns_topic.my_topic.arn
  description = "An output value for my block"
}
",
  "src/values.schema.json": "{
  "type": "object",
  "required": [],
  "properties": {
    "myValue": {
      "type": "string",
      "description": "An example of an input variable that will be used in the block"
    }
  }
}",
  "src/variables.tf": "variable "myValue" {
  description = "An example of an input variable that will be used in the block"
  type        = string
  default     = "my-default-value"
}
",
}
`;

exports[`tofu template 1`] = `
{
  "README.md": "# Bang Block

TODO: Add a description of the block or run \`kb enrich\` to generate one.",
  "kblock.yaml": "apiVersion: kblocks.io/v1
kind: Block
spec:
  engine: tofu
  definition:
    description: the bang bang description
    icon: ðŸ”¥
    readme: ./README.md
    schema: src/values.schema.json
    outputs:
      - myOutput
    group: bang.com
    version: v99
    kind: Bang
    plural: bangs
    color: red
    listKind: BangList
    shortNames:
      - b
      - ba
    categories:
      - cat1
      - cat2
    singular: bang
  operator:
    envSecrets:
      AWS_DEFAULT_REGION: aws-credentials
      AWS_ACCESS_KEY_ID: aws-credentials
      AWS_SECRET_ACCESS_KEY: aws-credentials
metadata:
  name: bangs.bang.com
",
  "src/main.tf": "# This is an example of an AWS SNS topic resource which will be created by the block
resource "aws_sns_topic" "my_topic" {
  name = var.myValue
}

# This is an example of an output value that will be returned by the block
output "myOutput" {
  value       = aws_sns_topic.my_topic.arn
  description = "An output value for my block"
}
",
  "src/values.schema.json": "{
  "type": "object",
  "required": [],
  "properties": {
    "myValue": {
      "type": "string",
      "description": "An example of an input variable that will be used in the block"
    }
  }
}",
  "src/variables.tf": "variable "myValue" {
  description = "An example of an input variable that will be used in the block"
  type        = string
  default     = "my-default-value"
}
",
}
`;

exports[`wing-k8s template 1`] = `
{
  "README.md": "# Bang Block

TODO: Add a description of the block or run \`kb enrich\` to generate one.",
  "kblock.yaml": "apiVersion: kblocks.io/v1
kind: Block
spec:
  engine: wing/k8s
  definition:
    outputs:
      - myOutput
    readme: ./README.md
    schema: ./src/block.w
    description: the bang bang description
    group: bang.com
    version: v99
    kind: Bang
    plural: bangs
    icon: ðŸ”¥
    color: red
    listKind: BangList
    shortNames:
      - b
      - ba
    categories:
      - cat1
      - cat2
    singular: bang
metadata:
  name: bangs.bang.com
",
  "src/block.w": "bring "cdk8s-plus-30" as k8s;

pub struct BangSpec {
  myInput: str;
}

pub class Bang {
  pub myOutput: str;

  new(spec: BangSpec) {

    let configMap = new k8s.ConfigMap(
      data: {
        "my-key": spec.myInput,
      }
    );

    this.myOutput = configMap.name;
  }
}",
  "src/package.json": "{
  "dependencies": {
    "@winglibs/k8s": "^0.0.8",
    "cdk8s-plus-30": "^2.0.6"
  }
}
",
}
`;
